flecsi/exec/hpx/task_prologue.hh

```diff
@@ -0,0 +1,223 @@
+      }
+      // if the task writes to the current argument then we must associate
+      // the future that represents the result of the task with the argument
+      std::lock_guard l(spinlock_pool::spinlock_for(&field_future));
+      field_future = future;
+      has_dependencies = true;
+    }
+
+    accessor.bind(storage);
+  } // visit generic topology
+
+public:
+  /*!
+    Delay the execution of the given task until all dependencies have been
+    satisfied.
+   */
+  template<typename R, typename Params, typename Task>
+  auto delay_execution(Params && params, std::string task_name, Task && task) {
+
+    if(dependencies.empty()) {
+      // TIP: param_buffers is an RAII type. We create an instance and give
+      // the object a reference to the parameters and name of the task. We
+      // then assign it to the `finalize` variable so it is not destroyed
+      // immediately. The object will be destroyed when reduce_internal()
+      // returns. The ~param_buffers() will then perform the necessary clean
+      // up on the parameters (mostly calling mutator.commit()).
+      auto finalize = param_buffers(params, task_name);
+
+      // directly execute task as no dependencies have to be taken into account
+      if constexpr(std::is_void_v<R>) {
+        task(std::forward<Params>(params), std::move(task_name));
+        return hpx::make_ready_future();
+      }
+      else {
+        return hpx::make_ready_future(
+          task(std::forward<Params>(params), std::move(task_name)));
+      }
+    }
+
+    // delay execution to start only after all dependencies have been satisfied
+    return ::hpx::dataflow(
+      ::hpx::launch::sync,
+      [&,
+        params = std::forward<Params>(params),
+        task_name = std::move(task_name)](auto && deps) mutable {
+        // TIP: param_buffers is an RAII type. We create an instance and give
+        // the object a reference to the parameters and name of the task. We
+        // then assign it to the `finalize` variable so it is not destroyed
+        // immediately. The object will be destroyed when reduce_internal()
+        // returns. The ~param_buffers() will then perform the necessary clean
+        // up on the parameters (mostly calling mutator.commit()).
+        auto finalize = param_buffers(params, task_name);
+
+        // rethrow exceptions
+        for(auto && f : deps)
+          f.get();
+
+        // invoke actual task
+        return task(std::move(params), std::move(task_name));
+      },
+      dependencies);
+  }
+
+  /*!
+    Make sure the future that is associated with the result of this
+    task is made ready once the task finishes executing.
+   */
+  template<typename T>
+  void attach_dependencies(::hpx::future<T> const & f) {
+    if(has_dependencies) {
+      flog_assert(future.valid(), "future must be valid");
+      hpx::traits::detail::get_shared_state(f)->set_on_completed(
+        [p = std::move(promise)]() mutable { p.set_value(); });
```

Comment:
This `std::move` tends to suggest that this function (or `delay_execution` if this stops being public) should be `&&`-qualified.  That might be a bit annoying for `reduce_internal`, but it really can be called only once&hellip;